### PR TITLE
Return the patch to the store from action.apply

### PR DESCRIPTION
**Pull request by sunesimonsen**
Merged `payload` and `apply` on actions and only accept cache patches at for actions entering the store.
The promise middleware takes care of supporting function and async function payloads and converts them into a cache patch.

This kind of change will make it a lot easier to create a logging middleware, action persistence and devtools.

Example logging middleware:

```js
store.useMiddleware(({ next, action }) => {
  console.log(action.payload);
  return next(action);
});
```

Screenshot 2021-01-30 at 02 22 51

### CODE REVIEW

**Comment by sunesimonsen** [comment]
I think I know how I can simplify things even more, so let me try that.

File: packages/nano-router-plugin/src/index.js [hunk 1]
```diff
@@ -6,13 +6,12 @@ export const nanoRouterPlugin = ({ routes, history }) => (store) => {
   const updateRouting = () => {
     store.dispatch({
       payload: {
-        route: router.route,
-        location: router.location,
-        params: router.params,
-        queryParams: router.queryParams,
-      },
-      apply: (cache, { payload }) => {
```
Comment (alexjeffburke) [inline packages/nano-router-plugin/src/index.js:14]: Ah, so apply is optional to care for situation where it's a static payload?

File: packages/preact/src/hooks/useObservable.spec.js [hunk 2]
```diff
@@ -23,8 +24,12 @@ const computedSum = {
 };
 
 const increment = () => ({
-  apply: (cache) => {
-    cache.update("(a|b)", (v) => v + 1);
```
Comment (alexjeffburke) [inline packages/preact/src/hooks/useObservable.spec.js:27]: This was that special update syntax you mentioned - the nee version is a little more verbose, but also very explicit. Do think the compactness was nice but think trading it away is a win. How did you feel about this?

File: packages/promise-middleware/src/index.js [hunk 3]
```diff
@@ -1,32 +1,18 @@
...
-      ...action,
-      payload: null,
-      error,
-    });
-
-    throw error;
-  }
+  return next({
+    ...action,
+    payload: await action.payload(readonlyCache(store.cache), ...args),
```
Comment (alexjeffburke) [inline packages/promise-middleware/src/index.js:16]: Was thinking exactly the same to disallow accidental writes when we discussed this!

File: packages/nano-router-plugin/src/index.js (hunk 1 above)
Comment (sunesimonsen) [reply packages/nano-router-plugin/src/index.js:14]: No apply is completely gone :-)

File: packages/preact/src/hooks/useObservable.spec.js (hunk 2 above)
Comment (sunesimonsen) [reply packages/preact/src/hooks/useObservable.spec.js:27]: I think it is an okay tradeoff.

**Pull request merged**